### PR TITLE
[AERIE-1880] ApplyWhen - scheduling goal windowing

### PR DESCRIPTION
**Pull request by pranav-super**
* **Tickets addressed:** AERIE-1880
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
This ticket was approached first by defining what it means for this change to be made (i.e. what it means for `ApplyWhen` to be applied on a goal), and then applying it in the backend goal by goal, and finally in the frontend. **NOTE: IN DEVELOPMENT, "ApplyWhen" WAS ALSO REFERRED TO AS "AnalyzeWhen", OWING TO A DISCUSSION THAT LED TO THE CREATION OF THIS TICKET - THEY PRESENTLY MEAN THE SAME THING.**

So, what does `ApplyWhen` mean? The idea is that an `Expression<Windows>` that, when evaluated, yields a `Windows`, can be given to restrict when a goal is active. This could be a strict temporal window (which needs to be wrapped as an expression - see the `WindowsWrapperExpression` class), or a resource window (which when simulated with the SimulationFacade yields a temporal window). **Do not that these windows can be disjoint!** What can be a little vague is what it means for each type of goal to have such a restriction applied:
- For a `CoexistenceGoal`, this means that any activities who match the coexistence criterion (the `forEach: ActivityExpression`) that lie outside of the window aren't considered for scheduling according to this goal. Similarly, if a matching activity is spliced by a set of goal windows, i.e. different windows contain parts of a matching activity, each separate window is treated as a match and an activity will be scheduled for each of those, following the goal's specification.
- For a `RecurrenceGoal`, this means recurrence only happens within this window. Note that recurrence is local, and partial recurrence isn't allowed. That means that every disjoint window has its own recurrence interval, and if a window cuts a recurrence interval short (i.e. the window is 11 seconds long but the recurrence interval is 2, meaning the last interval would be spliced), that interval won't be scheduled.
- Finally, for a `CardinalityGoal`, whatever can fit in the window is scheduled. 

The commits go in the order of tests, backend, then frontend. They start by detailing the creation of test files and supporting classes (such as a new sample `Goal` type under the `FooMissionModel`). 
Then the commits detail updates to the backend, first with clarifying an error message that appeared a few times in testing (it wasn't actually an error, more of a warning about the temporal network solver that places activities), then a refactor to use `Expression<Windows>`, a refactor of the goal types, and finally one of the EDSL itself. The EDSL update and the addition of its new tests are included as one commit. Finally, a boundary issue that was found in `RecurrenceGoal` is resolved in the last commit.

## Verification
The changes were validated through testing. This was primarily achieved by the `TestAnalyzeWhen.java` file, which aims to do extensive bounds checking for each of the goals to ensure that the implementation behaves consistently and predictably. There was some odd behavior encountered, namely with `UncontrollableDurationActivities` - this is documented.

Similarly, for the EDSL, both a parsing test was added (`SchedulingDSLCompilationTests.java`), as well as a test that verified parsing and the results of scheduling (`SchedulingIntegrationTests.java`). Only one of each was written, as one was sufficient to verify parsing, and one was sufficient to verify integration (as actual boundaries of the feature were tested in `TestAnalyzeWhen.java`).

## Documentation
The only possible invalidation would be regarding that of uncontrollable duration activities which were found to be problematic, but those were not documented. The documentation additions coming from this PR will be primarily just explaining how to use `ApplyWhen`, its edge cases, and the fact that `UncontrollableDurationActivities` have been found to be problematic.

## Future work
A new ticket was created to investigate `UncontrollableDurationActivities` - that is a direct product of the testing done while working on [this](https://jira.jpl.nasa.gov/projects/AERIE/issues/AERIE-1887) ticket.
Furthermore, in discussing this feature, a distinction was made between possible features, namely a `PlaceWhen` and `AnalyzeWhen` (of which this ticket handles the latter, despite analysis not being implemented) - a future feature might involve a separate `PlaceWhen` and `AnalyzeWhen`, as well as the creation of an analysis mode. For now, they can safely and reasonably be conflated, as analysis is not yet an existing feature.

### CODE REVIEW

File: scheduler-server/src/test/java/gov/nasa/jpl/aerie/scheduler/server/services/SchedulingIntegrationTests.java [hunk 1]
```diff
@@ -725,6 +725,56 @@ export default (): Goal => {
...
+                      window: Real.Resource("/plant").greaterThan(1.0)
+                  })""", true)
+        )
+    );
+
+    System.out.println("---RESULTS---");
+    for(MockMerlinService.PlannedActivityInstance p : results.updatedPlan()) {
+      System.out.println(p.startTime().toString() + ", " + p.type().toString());
+    }
+    System.out.println("-------------");
```
Comment (adrienmaillard) [inline scheduler-server/src/test/java/gov/nasa/jpl/aerie/scheduler/server/services/SchedulingIntegrationTests.java]: Remove the prints.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/TimeRangeExpression.java [hunk 2]
```diff
@@ -48,11 +48,12 @@ public Windows computeRange(final SimulationResults simulationResults, final Pla
       if(minTimepoint.isPresent() && maxTimepoint.isPresent()) {
         final var anchorActSearch = new ActivityExpression.Builder()
             .basedOn(actTemplate)
-            .startsIn(Window.between(
+            /*.startsIn(Window.between(
                 Duration.max(Duration.ZERO, minTimepoint.get()),
                 Window.Inclusivity.Inclusive,
                 maxTimepoint.get(),
-                Window.Inclusivity.Exclusive)).build();
+                Window.Inclusivity.Exclusive)).build();*/
```
Comment (adrienmaillard) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/TimeRangeExpression.java]: Remove commented lines.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ActivityTemplateGoal.java [hunk 3]
```diff
@@ -125,4 +127,11 @@ protected ActivityTemplateGoal() { }
    */
   protected ActivityCreationTemplate desiredActTemplate;
 
+  /**
+   * checked by getConflicts every time it is invoked to see if the Window(s)
+   * corresponding to when this goal has changed, which is unexpected behavior
+   * that needs to be caught
+   */
+  protected Windows evaluatedTemporalContext;
```
Comment (adrienmaillard) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ActivityTemplateGoal.java]: What about calling it `initialTemporalContext` or something that conveys the same meaning ?

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CardinalityGoal.java [hunk 4]
```diff
@@ -178,69 +178,90 @@ public boolean isUnsatisfiable(){
...
-    if (this.durationRange != null && !this.durationRange.contains(total)) {
-      if (total.compareTo(this.durationRange.start) < 0) {
-        durToSchedule = this.durationRange.start.minus(total);
-      } else if (total.compareTo(this.durationRange.end) > 0) {
-        logger.warn("Need to decrease duration of activities from the plan, impossible because scheduler cannot remove activities");
-        return List.of(new UnsatisfiableGoalConflict(this,
-                                                     "Need to decrease duration of activities from the plan, impossible because scheduler cannot remove activities"));
-      }
+    //make sure it hasn't changed
+    if (this.evaluatedTemporalContext != null && !windows.includes(this.evaluatedTemporalContext)) {
```
Comment (adrienmaillard) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CardinalityGoal.java]: Why use `includes()` and not `equals()` ?

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CoexistenceGoal.java [hunk 5]
```diff
@@ -157,10 +157,42 @@ protected CoexistenceGoal fill(CoexistenceGoal goal) {
...
-    final var conflicts = new java.util.LinkedList<Conflict>();
+  public java.util.Collection<Conflict> getConflicts(Plan plan, final SimulationResults simulationResults) { //TODO: check if window gets split and if so, notify user?
+
+    //NOTE: temporalContext IS A WINDOWS OVER WHICH THE GOAL APPLIES, USUALLY SOMETHING BROAD LIKE A MISSION PHASE
+    //NOTE: expr IS A WINDOWS OVER WHICH A COEXISTENCEGOAL APPLIES, FOR EXAMPLE THE WINDOWS CORRESPONDING TO 5 SECONDS AFTER EVERY BASICACTIVITY IS SCHEDULED
+    //NOTE: IF temporalContext IS SMALLER THAN expr OR SOMEHOW BISECTS IT, ODDS ARE THIS ISN'T ANTICIPATED USER BEHAVIOR. GENERALLY, ANALYZEWHEN SHOULDN'T BE PROVIDING
+    //        A SMALLER WINDOW, AND HONESTLY DOESN'T MAKE SENSE TO USE ON TOP BUT IS SUPPORTED TO MAKE CODE MORE CONSISTENT. IF ONE NEEDS TO USE ANALYZEWHEN ON TOP
+    //        OF COEXISTENCEGOAL THEY SHOULD PROBABLY REFACTOR THEIR COEXISTENCE GOAL. ONE SUCH USE WOULD BE IF THE COEXISTENCEGOAL WAS SPECIFIED IN TERMS OF
+    //        AN ACTIVITYEXPRESSION AND THEN ANALYZEWHEN WAS A MISSION PHASE, BUT I THINK ITS POSSIBLE TO JUST SPECIFY AN EXPRESSION<WINDOWS> THAT COMBINES THOSE.
```
Comment (adrienmaillard) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CoexistenceGoal.java]: Reformat: change to lowercase and remove "I" statement.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ProceduralCreationGoal.java [hunk 6]
```diff
@@ -200,16 +200,16 @@ protected ProceduralCreationGoal() { }
...
    */
-  private Collection<ActivityInstance> getRelevantGeneratedActivities(Plan plan) {
+  private Collection<ActivityInstance> getRelevantGeneratedActivities(Plan plan, SimulationResults simulationResults) {
 
     //run the generator in the plan context
     final var allActs = generator.apply(plan);
 
     //filter out acts that don't have a start time within the goal purview
-    final var goalContext = getTemporalContext();
+    final var goalContext = getTemporalContext(); //can extract Windows by calling evaluate, don't need to pass more to evaluate as it won't be another type of expression<windows>
```
Comment (adrienmaillard) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ProceduralCreationGoal.java]: Comment is not clear.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ProceduralCreationGoal.java [hunk 7]
```diff
@@ -200,16 +200,16 @@ protected ProceduralCreationGoal() { }
...
     final var filteredActs = allActs.stream().filter(
         act -> ((act.getStartTime() != null)
-                && goalContext.contains(act.getStartTime()))
+                && goalContext.evaluate(simulationResults).includesPoint(0,act.getStartTime()))
```
Comment (adrienmaillard) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ProceduralCreationGoal.java]: You evaluate the expression for each activities, I think this is inefficient. Evaluate once and pass the resulting windows.

File: scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java [hunk 8]
```diff
@@ -0,0 +1,1377 @@
...
+
+
+    problem.setGoals(List.of(goal));
+
+    final var solver = new PrioritySolver(problem);
+
+    var plan = solver.getNextSolution().orElseThrow();
+    for(ActivityInstance a : plan.getActivitiesByTime()){
+      System.out.println(a.getStartTime());
+    }
```
Comment (adrienmaillard) [inline scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java]: Remove or `logger.debug()`.

File: scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java [hunk 9]
```diff
@@ -0,0 +1,1377 @@
...
+        .build();
+
+
+    problem.setGoals(List.of(goal));
+
+    final var solver = new PrioritySolver(problem);
+
+    var plan = solver.getNextSolution().orElseThrow();
+    for(ActivityInstance a : plan.getActivitiesByTime()){
+      System.out.println(a.getStartTime());
```
Comment (adrienmaillard) [inline scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java]: Use `logger.debug` or remove. Same for similar prints below.

File: scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java [hunk 10]
```diff
@@ -0,0 +1,1377 @@
...
+    }
+
+    System.out.println("ahh!");
+
+    //assertTrue(TestUtility.activityStartingAtTime(plan,Duration.of(1, Duration.SECONDS), activityType)); //cutting off mid interval should fail
```
Comment (adrienmaillard) [inline scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java]: Remove commented lines.

File: scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java [hunk 11]
```diff
@@ -0,0 +1,1377 @@
...
+    problem.setGoals(List.of(goal));
+
+    final var solver = new PrioritySolver(problem);
+
+    var plan = solver.getNextSolution().orElseThrow();
+    for(ActivityInstance a : plan.getActivitiesByTime()){
+      System.out.println(a.getStartTime());
+    }
+
+    System.out.println("ahh!");
```
Comment (adrienmaillard) [inline scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java]: Remove print.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CardinalityGoal.java [hunk 12]
```diff
@@ -178,69 +178,90 @@ public boolean isUnsatisfiable(){
...
     final var conflicts = new LinkedList<Conflict>();
-    //at this point, have thrown exception if not satisfiable
-    //compute the missing association conflicts
-    for(final var act : acts){
-      if(!associatedActivitiesToThisGoal.contains(act) && planEvaluation.canAssociateMoreToCreatorOf(act)){
-        //they ALL have to be associated
-        conflicts.add(new MissingAssociationConflict(this, List.of(act)));
+
+    for(Window subWindow: windows) {
+      Windows timeDomain = this.expr.computeRange(simulationResults, plan, new Windows(subWindow));
```
Comment (adrienmaillard) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CardinalityGoal.java]: Why keep both `expr` and `temporalContext` now that they are the same kind ?

File: scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts [hunk 13]
```diff
@@ -162,6 +170,8 @@ declare global {
 
     public or(...others: Goal[]): Goal
 
+    public static applyWhen(opts: { goal: Goal, window: WindowsEDSL.Windows } ): Goal
```
Comment (adrienmaillard) [inline scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts]: Have you tested this directly in the UI ?

**Review by adrienmaillard — CHANGES_REQUESTED** [review]
Thank you for this work, you haven't started with the easiest of tasks. 

I was not there when this ticket was discussed so I still am unsure about the need for this feature. 

It seems to me that this new operator duplicates the goal effects. It is then some kind of composite goal. I would call it `applyForEach` or `forEach` in the edsl. `applyWhen` makes sense to me if it was a property of a goal, not a high level operator applying on goals. Opinions @mattdailis @patkenneally @JoelCourtney ? 

It is a departure from the original meaning of `temporalContext` that meant that we do not check/place activities elsewhere.  
 
There is clearly a difference between how it is presented in the edsl and how it is implemented in the backend. A more symmetric implementation (not that it would be better) would have been a new ForEach goal that would have actually created phantom goals with each small temporal context.

Anyway, I appreciate the thoroughness of your tests and the struggle that it must have been to deal with all the semantics at once.  

You'll see I have a few specific questions and minor remarks.

**Comment by mattdailis** [comment]
There was some confusion about whether ApplyWhen was available to users yet or not - I'm pasting the writeup below, feel free to edit this message, and then we can move it back into the wiki when we're ready:

## Restricting when a goal applied
The Aerie scheduler provides support for restricting _when_ a goal applies with the `Goal.applyWhen()` node. This node allows users to provide a `Window` (which could be a time or a resource-based window) and a goal to apply within that window.

The `Goal.applyWhen()` node, like the goals, takes one argument - an `options` argument, which requires two key-value pairs: a `goal` (which needs to be some declaration of a goal), and a window. What follows is an example that applies a daily recurrence goal only when a given resource is greater than 2. If the resource is less than two, then the goal is no longer applied.

image

#### Notes on boundaries:
* If you are trying to schedule an activity, or a recurrence within a window but that window cuts off either the activity or the recurrence interval (depending on the goal type), it will not be scheduled. For example, if you had a recurrence interval of 3 seconds, scheduling a 2 second activity each recurrence, and had the following window, you'd get the following:
```
RECURRENCE INTERVAL: [++-++-++-]
GOAL WINDOW:         [+++++----]
RESULT:              [++-------]
```
That, is, the second activity won't be scheduled as the goal window cuts off its recurrence interval. 
* Scheduling is _local_, not global. This means for every window that is matched (as it is possible to have disjoint windows, imagine a resource that fluctuates upward and downward but only applying that goal when the resource is over a certain value), the goal is applied individually. So, for that same recurrence interval setup as before, we could have:
```
RECURRENCE INTERVAL: [++-++-++-++-]
GOAL WINDOW:         [+++++--+++--]
RESULT:              [++-----++---] //(the second one is applied independently of the first!)
```
* When mapping out a temporal window to apply a goal over, keep in mind that the ending boundary of the goal is _exclusive_, i.e. if I want to apply a goal in the window of 10-12 seconds, it will apply only on seconds 10 and 11. This is in line with the [fencepost problem](https://icarus.cs.weber.edu/~dab/cs1410/textbook/3.Control/fencepost.html).

**Comment by pranav-super** [comment]
> It is a departure from the original meaning of temporalContext that meant that we do not check/place activities elsewhere.

By this do you mean the entire plan doesn't have activities placed elsewhere, or that activities aren't placed elsewhere with respect to the goal? Because in this case the latter is true.

> to me that this new operator duplicates the goal effects. It is then some kind of composite goal. I would call it applyForEach or forEach in the edsl. applyWhen makes sense to me if it was a property of a goal, not a high level operator applying on goals. 

> There is clearly a difference between how it is presented in the edsl and how it is implemented in the backend. A more symmetric implementation (not that it would be better) would have been a new ForEach goal that would have actually created phantom goals with each small temporal context.

The idea was to implement it as a property (decision made because temporalContext already existed and was referenced in a couple locations, so it made sense to leverage it, plus some sort of temporal restriction like this is a lot easier if implemented as a property of the goal), but to expose it to the edsl as a wrapper (I think this felt somewhat cleaner to represent). That being said, does anyone have strong opinions on implementing it as an optional goal property in the edsl @JoelCourtney @mattdailis @patkenneally?

> There was some confusion about whether ApplyWhen was available to users yet or not - I'm pasting the writeup below, feel free to edit this message, and then we can move it back into the wiki when we're ready:

My apologies

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ActivityTemplateGoal.java (hunk 3 above)
Comment (pranav-super) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ActivityTemplateGoal.java]: `initiallyEvaluatedTemporalContext` would be the best way to convey the meaning but I'm not sure if there's a more succinct way to say it (unless the fact that its a `Windows` object implies its been evaluated in which case your suggestion is the best name, I'm not sure what's the most readable option if one is unfamiliar with the code).

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CardinalityGoal.java (hunk 4 above)
Comment (pranav-super) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CardinalityGoal.java]: That's a good point, `equals()` is more accurate and more readable.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ProceduralCreationGoal.java (hunk 6 above)
Comment (pranav-super) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ProceduralCreationGoal.java]: The idea here was just to clarify the arguments being passed to `evaluate()`, as there were more arguments previously, and `getTemporalContext()` had a different return type (a Window). This clarification made more sense during the refactor but after everything is a lot more meaningless. It has been removed.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CoexistenceGoal.java (hunk 5 above)
Comment (pranav-super) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CoexistenceGoal.java]: Fixed

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CardinalityGoal.java (hunk 4 above)
Comment (pranav-super) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CardinalityGoal.java]: Fixed

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ActivityTemplateGoal.java (hunk 3 above)
Comment (pranav-super) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ActivityTemplateGoal.java]: Fixed

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/TimeRangeExpression.java (hunk 2 above)
Comment (pranav-super) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/TimeRangeExpression.java]: Fixed

File: scheduler-server/src/test/java/gov/nasa/jpl/aerie/scheduler/server/services/SchedulingIntegrationTests.java (hunk 1 above)
Comment (pranav-super) [reply scheduler-server/src/test/java/gov/nasa/jpl/aerie/scheduler/server/services/SchedulingIntegrationTests.java]: Fixed

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ProceduralCreationGoal.java (hunk 7 above)
Comment (pranav-super) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/ProceduralCreationGoal.java]: Fixed!

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CardinalityGoal.java (hunk 12 above)
Comment (pranav-super) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CardinalityGoal.java]: No reason, it has been factored out.

File: scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java (hunk 11 above)
Comment (pranav-super) [reply scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java]: Fixed

File: scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java (hunk 10 above)
Comment (pranav-super) [reply scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java]: Fixed

File: scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java (hunk 9 above)
Comment (pranav-super) [reply scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java]: Replaced all

File: scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java (hunk 8 above)
Comment (pranav-super) [reply scheduler/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java]: Fixed

File: scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts (hunk 13 above)
Comment (pranav-super) [reply scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts]: Yes, I was able to recreate the test from SchedulingIntegrationTests in the UI and it worked!
image
(I did restrict the window a bit in this one ^^)

**Comment by adrienmaillard** [comment]
Thank you for making the changes I have pointed out. I am still a little bit unsatisfied by the edsl implementation as a wrapper rather than a property. I'll wait for others to voice their opinions on this subject before approving.

**Comment by dyst5422** [comment]
I agree with @adrienmaillard. I think it feels weird to not have it be a method on a goal that modifies the context of that goal directly - especially since it doesn't modify the goal directly but returns a new goal. This is important because it allows us to still reuse goals in multiple contexts. IE
```typescript
const goal = Goal.ActivityRecurrenceGoal({
  activityTemplate: ActivityTemplates.BiteBanana({
    biteSize: 1
  }),
  interval: 24 * 60 * 60 * 1000 * 1000 //24 hours in microseconds
});
return goal.applyWhen(Real.Resource (Resources["/sample/resource/l"]).greaterThan(2.0))
.or(goal.applyWhen(Real.Resource (Resources["/sample/resource/l"]).lessThan(1.0)))
```

**Review by adrienmaillard — APPROVED** [review]
Thank you for this work.

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/WindowsWrapperExpression.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+  public WindowsWrapperExpression(final Windows windows) { this.windows = windows; }
+
+  @Override
+  public Windows evaluate(final SimulationResults results, final Window bounds, final Map<String, ActivityInstance> environment) {
+    return windows;
+  }
+
+  public String prettyPrint(final String prefix) {
+    return "WindowsWrapper: " + windows.toString();
+  }
```
Comment (JoelCourtney) [inline constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/WindowsWrapperExpression.java]: Please use the usual format that the other expressions use. e.g. from the Rate class:

```java
@Override
  public String prettyPrint(final String prefix) {
    return String.format(
        "(rate-of %s)",
        this.profile.prettyPrint(prefix + "  ")
    );
  }
```

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java [hunk 15]
```diff
@@ -32,32 +29,26 @@
...
+              microseconds -> Duration.of(microseconds, Duration.MICROSECONDS),
+              duration -> duration.in(Duration.MICROSECONDS)));
   private static final JsonParser<ClosedOpenInterval> intervalP =
       productP
           .field("start", durationP)
           .field("end", durationP)
           .map(Iso.of(
               untuple(ClosedOpenInterval::new),
               $ -> tuple($.start(), $.end())));
-
```
Comment (JoelCourtney) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java]: (not terribly important) lol why did you remove all the newlines in this file?

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java [hunk 16]
```diff
@@ -176,19 +168,19 @@ record CardinalityGoalDefinition(
     ) implements GoalSpecifier {}
     record GoalAnd(List<GoalSpecifier> goals) implements GoalSpecifier {}
     record GoalOr(List<GoalSpecifier> goals) implements GoalSpecifier {}
+    record GoalApplyWhen(
+        GoalSpecifier goal,
+        Expression<Windows> window
```
Comment (JoelCourtney) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java]: Please call it `windows` to avoid confusion

**Review by JoelCourtney — CHANGES_REQUESTED** [review]
Love it! I have one tiny change for the WindowsWrapperExpression class, but other than that I think its great.

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/WindowsWrapperExpression.java (hunk 14 above)
Comment (JoelCourtney) [reply constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/WindowsWrapperExpression.java]: Huh actually these functions are never used and the example I gave was wrong. Cool! Should be:

```java
@Override
  public String prettyPrint(final String prefix) {
    return String.format(
        "%s(rate-of %s)",
        prefix,
        this.profile.prettyPrint(prefix + "  ")
    );
  }
```

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java (hunk 15 above)
Comment (pranav-super) [reply scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java]: not sure, I guess just for formatting but I don't remember doing that often, might be a rebase quirk

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java (hunk 16 above)
Comment (pranav-super) [reply scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java]: Done!

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java (hunk 15 above)
Comment (pranav-super) [reply scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/models/SchedulingDSL.java]: Undone!

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/WindowsWrapperExpression.java (hunk 14 above)
Comment (pranav-super) [reply constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/WindowsWrapperExpression.java]: Done!

**Review by JoelCourtney — APPROVED** [review]
